### PR TITLE
Reorganized the golden tests

### LINKED ISSUE #755: Reorganize test suite
The test suite has grown organically over time, and it's become a bit of a mess. We should re-organize it in a more consistent and systematic fashion, also to avoid missing edge cases. See also https://github.com/well-typed/hs-bindgen/issues/726#issuecomment-2967410572 .

### PR DESCRIPTION
**Pull request by bolt12**
This PR closes #755 

* Reorganized `examples/golden/` into 11 subdirectories:
  * `types/`            - Type system tests (primitives, structs, unions, enums, typedefs)
  * `functions/`        - Function declarations, callbacks, attributes
  * `arrays/`           - Array tests
  * `macros/`           - C preprocessor macro tests
  * `declarations/`     - Scoping, redeclaration, linkage
  * `attributes/`       - GCC/Clang attributes
  * `program-analysis/` - Slicing, selection, reparsing
  * `documentation/`    - Doxygen, pragmas
  * `edge-cases/`       - UTF-8, asm, special names
  * `binding-specs/`    - Binding specification tests
  * `globals/`          - Global variable tests
  * `manual/`           - Manual examples (already existed)

### CODE REVIEW

File: hs-bindgen/examples/golden/attributes/asm.h
Comment (bolt12) [inline hs-bindgen/examples/golden/attributes/asm.h:1]: Move this to attributes.

File: hs-bindgen/examples/golden/macros/reparse.h
Comment (bolt12) [inline hs-bindgen/examples/golden/macros/reparse.h:1]: Move this to macros folder.

File: hs-bindgen/examples/golden/types/primitives/bool.h
Comment (bolt12) [inline hs-bindgen/examples/golden/types/primitives/bool.h:1]: In the types folder, further categorize the files: structs, enums, bool, etc..

File: hs-bindgen/test/hs-bindgen/Test/HsBindgen/Golden.hs [hunk 1]
```diff
@@ -82,59 +82,59 @@ testCases = manualTestCases ++ [
...
+    , defaultTest "types/type_qualifiers"
+    , defaultTest "types/unions"
+    , defaultTest "types/vector_test"
 
       --
       -- Clang diagnostics
       --
 
-    , testDiagnostic "attributes" $ \diag ->
+    , testDiagnostic "attributes/attributes" $ \diag ->
```
Comment (bolt12) [inline hs-bindgen/test/hs-bindgen/Test/HsBindgen/Golden.hs]: Have a global test list that is sorted by test name. In order to do that, associate these tests (`testDiagnostic`, e.g.) to variable names and add them to a list.

Comment (bolt12) [reply hs-bindgen/test/hs-bindgen/Test/HsBindgen/Golden.hs]: make them top level function definitions instead
